Implement closedby=any for dialog

lukewarlow opened the pull request:
#### ef0549626737a337f21297e9e0ddb2922b695e70
<pre>
Implement closedby=any for dialog
<a href="https://bugs.webkit.org/show_bug.cgi?id=297954">https://bugs.webkit.org/show_bug.cgi?id=297954</a>

Reviewed by Anne van Kesteren.

This adds an implementation for the closedby attribute&apos;s &quot;any&quot; value,
specifically covering the light dismiss behaviour.

Further additions will be needed to handle the close watcher aspect
of this attribute (e.g. none and closerequest values)

Canonical link: <a href="https://commits.webkit.org/309558@main">https://commits.webkit.org/309558@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/33a32a3f8e8a9fbcab352eb30957773cdca6da4d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/151027 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/23789 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/17359 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/159755 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/104463 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/152900 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/24220 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/24011 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/116600 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/82772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/153987 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/18722 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/135511 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/97321 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/17815 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/15765 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/7601 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/143011 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/127433 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/13428 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/162228 "Built successfully") | 
| [✅ 🛠 🧪 jsc-debug-arm64](https://ews-build.webkit.org/#/builders/171/builds/11826 "Built successfully and passed tests") | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/5353 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/14999 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/124607 "Passed tests") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/23591 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/19815 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/124795 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/33858 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/23581 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/135225 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/80000 "Built successfully") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/19860 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/11990 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/182636 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/23191 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/87482 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/46615 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/22903 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/23055 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/22957 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/dialog-closedby-bounds-clicking.html:
```diff
@@ -0,0 +1,100 @@
+<!DOCTYPE html>
```
These all need importing separately. They also need test_harness updated locally to add get_signal

nt1m replied:
The WPT sync landed!

lukewarlow commented on Source/WebCore/html/HTMLDialogElement.cpp:
```diff
@@ -282,6 +339,17 @@ void HTMLDialogElement::removedFromAncestor(RemovalType removalType, ContainerNo
 {
     HTMLElement::removedFromAncestor(removalType, oldParentOfRemovedTree);
     setIsModal(false);
+    if (document().settings().closedbyAttributeEnabled())
```
Some of these modifications need updating to account for the large number of edge cases that exist related to the open attribute

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10772,6 +10773,31 @@ HTMLElement* Document::topmostAutoPopover() const
+            && event.clientY() < dialogRect->bottom();
+        if (dialog->isOpen() && dialog->isModal() && !insideBounds)
+            return nullptr;
+    }
```
The HitTestRequest that populates the event target should have data on whether it's the backdrop or the dialog being clicked. I think it would be reasonable to extend the relevant Event interface with a non-web exposed field to pass on that data here

lukewarlow replied:
https://bugs.webkit.org/show_bug.cgi?id=298200 - it's not immediately clear to me how to trace the exact course of this PointerEvent, I've raised a follow-up issue to look into this.

annevk commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10765,6 +10766,31 @@ HTMLElement* Document::topmostAutoPopover() const
+HTMLDialogElement* Document::nearestClickedDialog(const PointerEvent& event, Node& target) const
+{
+    if (RefPtr dialog = dynamicDowncast<HTMLDialogElement>(target)) {
+        Ref<DOMRect> dialogRect = dialog->getBoundingClientRect();
```
```suggestion
        Ref dialogRect = dialog->getBoundingClientRect();
```

annevk commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10765,6 +10766,31 @@ HTMLElement* Document::topmostAutoPopover() const
+            return nullptr;
+    }
+
+    auto currentNode = dynamicDowncast<Element>(target);
```
RefPtr

annevk commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10765,6 +10766,31 @@ HTMLElement* Document::topmostAutoPopover() const
+    auto currentNode = dynamicDowncast<Element>(target);
+
+    while (currentNode) {
+        if (RefPtr dialog = dynamicDowncast<HTMLDialogElement>(currentNode)) {
```
This can be *currentNode, no? No need for a null check.

lukewarlow replied:
I'm not sure I fully understand this?

annevk replied:
Since we are inside the while loop currentNode is non-null. So we should dereference so dynamicDowncast doesn’t have to do a redundant null check.

lukewarlow replied:
Ah in that case, I believe that's done.

annevk commented on Source/WebCore/html/HTMLDialogElement.cpp:
```diff
@@ -58,6 +59,49 @@ HTMLDialogElement::HTMLDialogElement(const QualifiedName& tagName, Document& doc
+        return ClosedByState::None;
+    }
+
+    return closedByState();
```
Seems better to capture the return value in a variable so we don't have multiple attribute calls?

annevk commented on Source/WebCore/html/HTMLDialogElement.cpp:
```diff
@@ -58,6 +59,49 @@ HTMLDialogElement::HTMLDialogElement(const QualifiedName& tagName, Document& doc
+ClosedByState HTMLDialogElement::computedClosedByState() const
+{
+    if (closedByState() == ClosedByState::Auto) {
+        if (m_isModal)
```
`return m_isModal ? ... : ...;`

annevk commented on Source/WebCore/html/HTMLDialogElement.cpp:
```diff
@@ -278,12 +322,65 @@ bool HTMLDialogElement::supportsFocus() const
+{
+    HTMLElement::didFinishInsertingNode();
+    Ref document = this->document();
+    if (document->settings().closedbyAttributeEnabled()) {
```
We want this check in `insertedIntoAncestor` instead I think for now and assert it's true in this callback. That way we do less work overall.

annevk commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10895,6 +10896,33 @@ HTMLElement* Document::topmostAutoPopover() const
+
+    RefPtr currentNode = dynamicDowncast<Element>(target);
+
+    RefPtr<HTMLDialogElement> clickedDialog;
```
This does not appear to be used.

annevk commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10895,6 +10896,33 @@ HTMLElement* Document::topmostAutoPopover() const
+            return nullptr;
+    }
+
+    RefPtr currentNode = dynamicDowncast<Element>(target);
```
I think in general this could be slightly more efficient. We could this upfront. Then we could dynamic downcast the result of this to dialog. And if we used do/while below we could even reuse that dialog in that loop so we need one less downcast.

lukewarlow replied:
I'm not sure I fully understand this.

> We could this upfront.

Can you explain what this was meant to say that might help me.

I'm guessing somethinglike below but with the dialog downcast pulled out too somehow?

```
    RefPtr currentNode = dynamicDowncast<Element>(target);
    if (!currentNode)
        return nullptr;

    do {
        if (RefPtr dialog = dynamicDowncast<HTMLDialogElement>(*currentNode)) {
            if (dialog->isOpen())
                return dialog;
        }
        currentNode = currentNode->parentElementInComposedTree();
    } while (currentNode);

    return nullptr;
    ```

annevk replied:
I was thinking moving the `currentNode` all the way to the top of the function. And then making `dialog` global for the function scope as well. And then adjust the code as needed. Does that help?

lukewarlow replied:
```c++
RefPtr<HTMLDialogElement> Document::nearestClickedDialog(const PointerEvent& event, Node& target) const
{
    RefPtr currentNode = dynamicDowncast<Element>(target);
    RefPtr dialog = dynamicDowncast<HTMLDialogElement>(target);

    if (dialog) {
        Ref dialogRect = dialog->getBoundingClientRect();
        auto insideBounds = event.clientX() >= dialogRect->left()
            && event.clientX() < dialogRect->right()
            && event.clientY() >= dialogRect->top()
            && event.clientY() < dialogRect->bottom();
        if (dialog->isOpen() && dialog->isModal() && !insideBounds)
            return nullptr;
    }

    do {
        if ((dialog = dynamicDowncast<HTMLDialogElement>(*currentNode))) {
            if (dialog->isOpen())
                return dialog;
        }
        currentNode = currentNode->parentElementInComposedTree();
    } while (currentNode);

    return nullptr;
}
```

So to confirm something like this? (sorry if I'm missing something obvious)

annevk replied:
Yeah, in do/while you only need to assign dialog after you assign currentNode because initially it should have that value already.

annevk replied:
Oh also when you assign dialog initially you want to pass in currentNode. That should require less casting.

annevk commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -11000,6 +11028,41 @@ void Document::handlePopoverLightDismiss(const PointerEvent& event, Node& target
+
+    ASSERT(event.type() == eventNames().pointerupEvent);
+
+    auto sameTarget = ancestor.get() == m_dialogPointerDownTarget;
```
```suggestion
    bool sameTarget = ancestor.get() == m_dialogPointerDownTarget;
```
We don't use auto for bools.

annevk reviewed (approved):
This looks okay. Not entirely sure about using CommonAtomStrings for this, but it does seem like there is precedent...

lukewarlow committed:
Implement closedby=any for dialog
https://bugs.webkit.org/show_bug.cgi?id=297954

Reviewed by Anne van Kesteren.

This adds an implementation for the closedby attribute's "any" value,
specifically covering the light dismiss behaviour.

Further additions will be needed to handle the close watcher aspect
of this attribute (e.g. none and closerequest values)

Canonical link: https://commits.webkit.org/309558@main
